20240912-R-scheduler-auto-scale.md

```diff
@@ -0,0 +1,110 @@
+# Auto-Scale the Scheduler service
+
+* Author(s): Cassie Coyle (@cicoyle)
+* State: Review
+* Updated: 2024-09-12
+
+## Overview
+
+This proposal outlines a design for auto-scaling the Schedulers, which is needed for the Scheduler service to be Stable.
+The Scheduler's internal cron scheduler library code will need to be refactored to be dynamic. Runtime will need to be
+updated to also account for dynamically connecting to Schedulers when they are scaled up or down. The Scheduler server
+code will need to be updated such that it optionally runs embedded etcd on different Scheduler instances.
+
+
+## Background
+
+In Dapr v1.14 we introduced the Scheduler control plane service to support the Jobs API along with scaling the Actor
+reminder system and the Workflow API (since it uses Actor reminders under the hood). The Scheduler encompasses running
+itself as a new process, an embedded etcd instance, and an internal cron scheduler. There are a few things needed to
+make the Scheduler a stable control plane service. This proposal will be detailing how to support the auto-scaling of
+Schedulers and secondly enhancing the Schedulers such that we optionally run etcd instances in the Schedulers under
+certain circumstances that make sense. When scaling beyond 5 etcd instances, there are little performance improvements, 
+so it makes sense for Schedulers to become stateless at this point. For example, we could have 20 Schedulers, but only
+5 of them would contain the embedded etcd.
```

Comment:
I would say 3. Even 5 does not give much. 3 is good because it is the minimum number of replicas for HA, giving quorum vote even when one etcd instance is down. 5 means 2 etcd instances can be down. The higher availability comes at a cost of running with 5 schedulers at minimum. So, 3 is cheaper for auto-scaling scenarios.

Reply:
Another option is for user to pick which mode they want?
- non-HA: 1 etcd instance
- HA: 3 etcd instances
- "super HA": 5 etcd instances

Reply:
For consistency sake, and if 5 doesn't offer much more benefit than 3, then Im fine with 3.